velox/connectors/hive/HivePartitionName.h

```diff
@@ -0,0 +1,177 @@
+  /// Returns a vector of (key, value) pairs for all partition columns.
+  /// @tparam Fn A callable that converts a value to a partition
+  ///         string. Takes (value, type, columnIndex) and returns string.
+  /// @param partitionId The partition ID (row index) to extract values from.
+  /// @param partitionValues RowVector containing partition values.
+  /// @param nullValueString The string to use for null values.
+  /// @param toPartitionName Callable to convert the value to a string.
+  template <typename Fn>
+  static std::vector<std::pair<std::string, std::string>> partitionKeyValues(
+      uint32_t partitionId,
+      const RowVectorPtr& partitionValues,
+      const std::string& nullValueString,
+      const Fn& toPartitionName);
+
+  /// Generate a Hive partition directory name from partition values for
+  /// partitionId.
+  ///
+  /// @param partitionId The row index in partitionValues to extract values
+  ///        from.
+  /// @param partitionValues RowVector containing partition values. Each
+  ///        child vector represents a partition column, and the row at
+  ///        partitionId contains the values for this partition.
+  /// @param partitionKeyAsLowerCase Controls whether partition column names
+  ///        should be converted to lowercase in the output. When true, column
+  ///        names are lowercased (e.g., "year=2025"); when false, original
+  ///        casing is preserved (e.g., "Year=2025").
+  /// @return A formatted partition directory name string. Null values are
+  /// represented as
+  ///         __HIVE_DEFAULT_PARTITION__.
+  static std::string partitionName(
+      uint32_t partitionId,
+      const RowVectorPtr& partitionValues,
+      bool partitionKeyAsLowerCase);
+};
+
+namespace {
+
+// Unified template function to extract partition key-value string from a
+// vector. Used by both Hive and Iceberg partition name generators.
+//
+// @tparam Kind The TypeKind of the partition column.
+// @tparam Fn A callable that converts a value to a partition string.
+// @param partitionVector The vector containing partition values.
+// @param row The row index to extract the value from.
+// @param name The partition column name.
+// @param type The type of the partition column.
+// @param columnIndex The column index in the partition values.
+// @param nullValue The string to use for null values ("" for Hive, "null" for
+//        Iceberg).
+// @param toPartitionName Callable to convert the value to a string.
+// @return A pair of (column_name, formatted_value).
+template <TypeKind Kind, typename Fn>
+std::pair<std::string, std::string> makePartitionKeyValueString(
+    const BaseVector& partitionVector,
+    vector_size_t row,
+    const std::string& name,
+    const TypePtr& type,
+    int columnIndex,
+    const std::string& nullValue,
+    const Fn& toPartitionName) {
+  using T = typename TypeTraits<Kind>::NativeType;
+  if (partitionVector.as<SimpleVector<T>>()->isNullAt(row)) {
```

Comment:
This check can happen before calling the template. Doing this will help simplify this function further by removing 'nullValue' parameter.